exp/controllers/gcpmanagedcontrolplane_controller.go

```diff
@@ -29,33 +50,142 @@ import (
+	}
+
+	if err = c.Watch(
+		&source.Kind{Type: &clusterv1.Cluster{}},
+		handler.EnqueueRequestsFromMapFunc(util.ClusterToInfrastructureMapFunc(ctx, infrav1exp.GroupVersion.WithKind("GCPManagedCluster"), mgr.GetClient(), &infrav1exp.GCPManagedControlPlane{})),
+		predicates.ClusterUnpausedAndInfrastructureReady(log),
+	); err != nil {
+		return fmt.Errorf("failed adding a watch for ready clusters: %w", err)
+	}
+
+	return nil
+}
+
+func (r *GCPManagedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultedLoopTimeout(r.ReconcileTimeout))
+	defer cancel()
+
+	log := ctrl.LoggerFrom(ctx)
+
+	// Get the control plane instance
+	gcpManagedControlPlane := &infrav1exp.GCPManagedControlPlane{}
+	if err := r.Client.Get(ctx, req.NamespacedName, gcpManagedControlPlane); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Get the cluster
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, gcpManagedControlPlane.ObjectMeta)
+	if err != nil {
+		log.Error(err, "Failed to retrieve owner Cluster from the API Server")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info("Cluster Controller has not yet set OwnerRef")
+		return ctrl.Result{}, nil
+	}
+
+	if annotations.IsPaused(cluster, gcpManagedControlPlane) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	managedControlPlaneScope, err := scope.NewManagedControlPlaneScope(scope.ManagedControlPlaneScopeParams{
+		Client:                 r.Client,
+		Cluster:                cluster,
+		GCPManagedControlPlane: gcpManagedControlPlane,
+	})
+	if err != nil {
+		return ctrl.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
 
-	// TODO(user): your logic here
+	// Always close the scope when exiting this function so we can persist any GCPMachine changes.
+	defer func() {
+		if err := managedControlPlaneScope.Close(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !gcpManagedControlPlane.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, managedControlPlaneScope)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcile(ctx, managedControlPlaneScope)
+}
+
+func (r *GCPManagedControlPlaneReconciler) reconcile(ctx context.Context, managedControlPlaneScope *scope.ManagedControlPlaneScope) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.Info("Reconciling GCPManagedControlPlane")
+
+	controllerutil.AddFinalizer(managedControlPlaneScope.GCPManagedControlPlane, infrav1.ClusterFinalizer)
+	if err := managedControlPlaneScope.PatchObject(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	reconcilers := []cloud.Reconciler{
+		clusters.New(managedControlPlaneScope),
+	}
+
+	for _, r := range reconcilers {
+		if err := r.Reconcile(ctx); err != nil {
```

Comment:
This will need to be changed if we support requeue from the reconcilers.

Reply:
done